crates/engine/tree/src/tree/root.rs

```diff
@@ -118,31 +218,389 @@ where
 
     /// Handles state updates.
     fn on_state_update(
-        _view: &reth_provider::providers::ConsistentDbView<impl Send + 'static>,
-        _input: &std::sync::Arc<reth_trie::TrieInput>,
-        _state: EvmState,
+        view: ConsistentDbView<Factory>,
+        input: Arc<TrieInput>,
+        update: EvmState,
+        state: &mut HashedPostState,
+        proof_sequence_number: u64,
+        state_root_message_sender: Sender<StateRootMessage>,
     ) {
-        // Default implementation of state update handling
-        // TODO: calculate hashed state update and dispatch proof gathering for it.
+        let mut hashed_state_update = HashedPostState::default();
+        for (address, account) in update {
+            if account.is_touched() {
+                let hashed_address = keccak256(address);
+
+                let destroyed = account.is_selfdestructed();
+                hashed_state_update.accounts.insert(
+                    hashed_address,
+                    if destroyed || account.is_empty() { None } else { Some(account.info.into()) },
+                );
+
+                if destroyed || !account.storage.is_empty() {
+                    let storage = account.storage.into_iter().filter_map(|(slot, value)| {
+                        value
+                            .is_changed()
+                            .then(|| (keccak256(B256::from(slot)), value.present_value))
+                    });
+                    hashed_state_update
+                        .storages
+                        .insert(hashed_address, HashedStorage::from_iter(destroyed, storage));
+                }
+            }
+        }
+
+        // Dispatch proof gathering for this state update
+        let targets = hashed_state_update
+            .accounts
+            .keys()
+            .filter(|hashed_address| {
+                !state.accounts.contains_key(*hashed_address) &&
+                    !state.storages.contains_key(*hashed_address)
+            })
+            .map(|hashed_address| (*hashed_address, HashSet::default()))
+            .chain(hashed_state_update.storages.iter().map(|(hashed_address, storage)| {
+                (*hashed_address, storage.storage.keys().copied().collect())
+            }))
+            .collect::<HashMap<_, _>>();
+
+        rayon::spawn(move || {
+            let provider = match view.provider_ro() {
+                Ok(provider) => provider,
+                Err(error) => {
+                    error!(target: "engine::root", ?error, "Could not get provider");
+                    return;
+                }
+            };
+
+            // TODO: replace with parallel proof
+            let result =
+                Proof::overlay_multiproof(provider.tx_ref(), input.as_ref().clone(), targets);
```

Comment:
this clone is expensive, we should find a way to get rid of it

Reply:
it happens inside the `ParallelProof` anyway, but maybe we can avoid it altogether https://github.com/paradigmxyz/reth/blob/e8d63e4a0b060e1f8327059da78b5d0941aa6733/crates/trie/parallel/src/proof.rs#L65-L69